packages/sw-precaching/src/lib/controllers/base-manager.js

```diff
@@ -0,0 +1,171 @@
+   * the next install event.
+   */
+  _addEntryToInstallList(precacheEntry, isRevisioned) {
+    const entryID = precacheEntry.entryID;
+    const previousEntry = this._entriesToCache[precacheEntry.entryID];
+
+    if (!previousEntry) {
+      // This entry isn't in the install list
+      this._entriesToCache[entryID] = precacheEntry;
+      return;
+    }
+
+    this._onDuplicateEntryFound(precacheEntry, previousEntry);
+  }
+
+  /**
+   * This method manages the actual install event to cache the revisioned
+   * assets.
+   * @return {Promise} The promise resolves when all the desired assets are
+   * cached.
+   */
+  async _performInstallStep(cacheName) {
+    if (Object.keys(this._entriesToCache).length === 0) {
+      return;
+    }
+
+    let openCache = await this._getCache();
+
+    const entriesToCache = Object.values(this._entriesToCache);
+    const cachePromises = entriesToCache.map(async (precacheEntry) => {
+      return this._cacheEntry(precacheEntry, openCache);
+    });
+
+    // Wait for all requests to be cached.
+    await Promise.all(cachePromises);
+  }
+
+  async _cacheEntry(precacheEntry, openCache) {
+    const isCached = await this._isAlreadyCached(precacheEntry, openCache);
+    if (isCached) {
+      return;
+    }
+
+    let response = await fetch(precacheEntry.getNetworkRequest(), {
+        credentials: 'same-origin',
+      });
+
+    if (response.ok) {
+      await openCache.put(precacheEntry.request, response);
+
+      await this._onEntryCached(precacheEntry);
+    } else {
+      throw ErrorFactory.createError('request-not-cached', {
+        message: `Failed to get a cacheable response for ` +
+          `'${precacheEntry.request.url}'`,
+      });
+    }
+  }
+
+  /**
+   * Once the install event has occured and the previous entries need
+   * to be deleted from the cache, this method will compare the URL's
+   * and figure out which assets are no longer required to be precached.
+   */
+  async _cleanUpOldEntries() {
+    const requestsCachedOnInstall = Object.values(this._entriesToCache)
+      .map((entry) => entry.request.url);
+
+    let openCache = await this._getCache();
+    const allCachedRequests = await openCache.keys();
+
+    const cachedRequestsToDelete = allCachedRequests.filter((cachedRequest) => {
+      if (requestsCachedOnInstall.includes(cachedRequest.url)) {
+        return false;
+      }
+      return true;
+    });
+
+    await Promise.all(
+      cachedRequestsToDelete.map((cachedRequest) => {
+        return openCache.delete(cachedRequest);
+      })
+    );
+  }
+
+  _onDuplicateEntryFound(newEntry, previous) {
+    throw new Error('This should be overriden by extending class.');
+  }
+
+  _getCache() {
+    throw new Error('This should be overriden by extending class.');
+  }
+
+  _isAlreadyCached(precacheEntry, openCache) {
+    throw new Error('This should be overriden by extending class.');
+  }
+
+  _onEntryCached(precacheEntry) {
+    // NOOP
+  }
```

Comment:
This can throw and require both extending classes implement this (one of which will be a noop and one will be doing same as it is currently) or we can NOOP here and leave one class implementing and the other not.

Happy to change if you feel strongly.